File: examples/python-repo/ecommerce/application/payment/process_payment_handler.py
from ecommerce.application.payment.payment_dto import PaymentDto
from ecommerce.domain.payment.payment import Payment


class ProcessPaymentHandler:
    def __init__(self, payment_repo, order_repo):
        self._payment_repo = payment_repo
        self._order_repo = order_repo

    def execute(self, order_id, payment_method):
        order = self._order_repo.find_by_id(order_id)
        if order is None:
            raise ValueError(f"order not found: {order_id}")

        if not payment_method or not payment_method.strip():
            raise ValueError("payment method must not be blank")

        existing = self._payment_repo.find_by_order_id(order_id)
        if existing is not None:
            raise RuntimeError(f"payment already exists for order: {order_id}")

        payment_id = self._payment_repo.next_id()
        payment = Payment(payment_id, order.id, order.total(), payment_method)

        payment.process()

        # DELIBERATE: deeply nested ifs (complexity)
        if payment.amount is not None:
            if not payment.amount.is_negative():
                if payment.method == "credit_card":
                    if payment.amount.amount > 0:
                        success = True
                    else:
                        success = False
                else:
                    success = True
            else:
                success = False
        else:
            success = False

        if success:
            payment.complete()
            order.mark_paid()
            self._order_repo.save(order)
        else:
            payment.fail()

        self._payment_repo.save(payment)
        return PaymentDto.from_domain(payment)

File: examples/java-repo/src/main/java/com/example/ecommerce/application/payment/ProcessPaymentHandler.java
package com.example.ecommerce.application.payment;

import com.example.ecommerce.domain.order.Order;
import com.example.ecommerce.domain.order.OrderRepository;
import com.example.ecommerce.domain.payment.Payment;
import com.example.ecommerce.domain.payment.PaymentRepository;
import java.util.Optional;

public class ProcessPaymentHandler {

    private final PaymentRepository paymentRepo;
    private final OrderRepository orderRepo;

    public ProcessPaymentHandler(PaymentRepository paymentRepo, OrderRepository orderRepo) {
        this.paymentRepo = paymentRepo;
        this.orderRepo = orderRepo;
    }

    // DELIBERATE: duplicated validation block with PlaceOrderHandler (CPD)
    public PaymentDto execute(long orderId, String paymentMethod) {
        Optional<Order> orderOpt = orderRepo.findById(orderId);
        if (!orderOpt.isPresent()) {
            throw new IllegalArgumentException("order not found: " + orderId);
        }
        Order order = orderOpt.get();

        if (paymentMethod == null || paymentMethod.isBlank()) {
            throw new IllegalArgumentException("payment method must not be blank");
        }

        Optional<Payment> existingPayment = paymentRepo.findByOrderId(orderId);
        if (existingPayment.isPresent()) {
            throw new IllegalStateException("payment already exists for order: " + orderId);
        }

        long paymentId = paymentRepo.nextId();
        Payment payment = new Payment(paymentId, order.getId(), order.total(), paymentMethod);

        payment.process();

        boolean success = processWithGateway(payment);
        if (success) {
            payment.complete();
            order.markPaid();
            orderRepo.save(order);
        } else {
            payment.fail();
        }

        paymentRepo.save(payment);
        return PaymentDto.fromDomain(payment);
    }

    private boolean processWithGateway(Payment payment) {
        return !payment.getAmount().isNegative();
    }
}
